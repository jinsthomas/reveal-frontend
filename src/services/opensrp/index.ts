import { OpenSRPService as OpenSRPServiceWeb } from '@opensrp/server-service';
import { IncomingHttpHeaders } from 'http';
import { OPENSRP_API_BASE_URL } from '../../configs/env';
import store from '../../store';
import { getAccessToken } from '../../store/selectors';

/** allowed http methods */
type HTTPMethod = 'GET' | 'POST' | 'PUT' | 'DELETE';

/** get default HTTP headers for OpenSRP service */
export function getDefaultHeaders(
  accept: string = 'application/json',
  authorizationType: string = 'Bearer',
  contentType: string = 'application/json;charset=UTF-8'
): IncomingHttpHeaders {
  return {
    accept,
    authorization: `${authorizationType} ${getAccessToken(store.getState())}`,
    'content-type': contentType,
  };
}

/** interface to describe URL params object */
export interface URLParams {
  [key: string]: string | number | boolean;
}

/** converts filter params object to string
 * @param {URLParams} obj - the object representing filter params
 * @returns {string} filter params as a string
 */
export function getFilterParams(obj: URLParams | {}): string {
  return Object.entries(obj)
    .map(([key, val]) => `${key}:${val}`)
    .join(',');
}

/** get payload for fetch
 * @param {AbortSignal} signal - signal object that allows you to communicate with a DOM request
 * @param {HTTPMethod} method - the HTTP method
 * @returns the payload
 */

export function getPayloadOptions(_: AbortSignal, method: HTTPMethod) {
  return {
    headers: getDefaultHeaders() as HeadersInit,
    method,
  };
}

/** The OpenSRP service class
 * Extends class OpenSRPService from OpenSRPService web
 * Sample usage:
 * -------------
 * const service = new OpenSRPService('the-endpoint');
 *
 * **To list all objects**: service.list()
 *
 * **To get one object**: service.read('the-object-identifier')
 *
 * **To create a new object**: service.create(theObject)
 *
 * **To update an object**: service.update(theObject)
 */
export class OpenSRPService extends OpenSRPServiceWeb {
  constructor(
    endpoint: string,
    baseURL: string = OPENSRP_API_BASE_URL,
    getPayload: typeof getPayloadOptions = getPayloadOptions
  ) {
    super(baseURL, endpoint, getPayload);
  }
  /** create method
   * Send a POST request to the general endpoint containing the new object data
   * Successful requests will result in a HTTP status 201 response with no body
   * @param {T} data - the data to be POSTed
   * @param {params} params - the url params object
   * @param {HTTPMethod} method - the HTTP method
   * @returns the object returned by API
   */
  public async create<T>(data: T, params: any = null, method: HTTPMethod = 'POST'): Promise<{}> {
    const url = OpenSRPService.getURL(this.generalURL, params);
    const payload = {
      ...this.getOptions(this.signal, method),
      'Cache-Control': 'no-cache',
      Pragma: 'no-cache',
      body: JSON.stringify(data),
    };
    const response = await fetch(url, payload);

    if (!response.ok || response.status !== 201) {
      throw new Error(
        `OpenSRPService create on ${this.endpoint} failed, HTTP status ${response.status}`
      );
    }

    return {};
  }
}
