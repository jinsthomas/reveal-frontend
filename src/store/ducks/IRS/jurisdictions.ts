import { get, keyBy, values } from 'lodash';
import { AnyAction, Store } from 'redux';
import SeamlessImmutable from 'seamless-immutable';

/** the reducer name */
export const reducerName = 'IRSJurisdictions';

/** IRSJurisdiction interface */
export interface IRSJurisdiction {
  id: string;
  jurisdiction_depth: number;
  jurisdiction_id: string;
  jurisdiction_name: string;
  jurisdiction_name_path: string[];
  jurisdiction_parent_id: string;
  jurisdiction_path: string[];
  plan_id: string;
}

export interface NamibiaIRSJurisdiction extends IRSJurisdiction {
  lockedfirst: number;
  lockedmopup: number;
  refusalsfirst: number;
  refusalsmopup: number;
  sprayeffectiveness: number;
  structuresfound: number;
  structuressprayed: number;
  targetcoverage: number;
  target_2019: number;
}

// actions

/** IRS_JURISDICTION_FETCHED action type */
export const IRS_JURISDICTIONS_FETCHED =
  'reveal/reducer/IRS/IRSJurisdiction/IRS_JURISDICTIONS_FETCHED';

/** IRS_JURISDICTION_FETCHED action type */
export const REMOVE_IRS_JURISDICTIONS =
  'reveal/reducer/IRS/IRSJurisdiction/REMOVE_IRS_JURISDICTIONS';

/** IRS_JURISDICTION_FETCHED action type */
export const ADD_IRS_JURISDICTION = 'reveal/reducer/IRS/IRSJurisdiction/ADD_IRS_JURISDICTION';

/** interface for fetch IRSJurisdictions action */
interface FetchIRSJurisdictionsAction extends AnyAction {
  IRSJurisdictionsById: { [key: string]: NamibiaIRSJurisdiction };
  type: typeof IRS_JURISDICTIONS_FETCHED;
}

/** interface for removing IRSJurisdictions action */
interface RemoveIRSJurisdictionsAction extends AnyAction {
  IRSJurisdictionsById: { [key: string]: NamibiaIRSJurisdiction };
  type: typeof REMOVE_IRS_JURISDICTIONS;
}

/** interface for adding a single IRSJurisdictions action */
interface AddIRSJurisdictionAction extends AnyAction {
  obj: NamibiaIRSJurisdiction;
  type: typeof ADD_IRS_JURISDICTION;
}

/** Create type for IRSJurisdiction reducer actions */
export type IRSJurisdictionActionTypes =
  | FetchIRSJurisdictionsAction
  | RemoveIRSJurisdictionsAction
  | AddIRSJurisdictionAction
  | AnyAction;

// action creators

/**
 * Fetch Plan Definitions action creator
 * @param {IRSJurisdiction[]} objList - list of plan definition objects
 */
export const fetchIRSJurisdictions = (
  objList: NamibiaIRSJurisdiction[] = []
): FetchIRSJurisdictionsAction => ({
  IRSJurisdictionsById: keyBy(objList, 'id'),
  type: IRS_JURISDICTIONS_FETCHED,
});

/** Reset plan definitions state action creator */
export const removeIRSJurisdictions = () => ({
  IRSJurisdictionsById: {},
  type: REMOVE_IRS_JURISDICTIONS,
});

/**
 * Add one Plan Definition action creator
 * @param {IRSJurisdiction} obj - the plan definition object
 */
export const addIRSJurisdiction = (obj: NamibiaIRSJurisdiction): AddIRSJurisdictionAction => ({
  obj,
  type: ADD_IRS_JURISDICTION,
});

// the reducer

/** interface for IRSJurisdiction state */
interface IRSJurisdictionState {
  IRSJurisdictionsById: { [key: string]: NamibiaIRSJurisdiction };
}

/** immutable IRSJurisdiction state */
export type ImmutableIRSJurisdictionState = IRSJurisdictionState &
  SeamlessImmutable.ImmutableObject<IRSJurisdictionState>;

/** initial IRSJurisdiction state */
const initialState: ImmutableIRSJurisdictionState = SeamlessImmutable({
  IRSJurisdictionsById: {},
});

/** the IRSJurisdiction reducer function */
export default function reducer(
  state = initialState,
  action: IRSJurisdictionActionTypes
): ImmutableIRSJurisdictionState {
  switch (action.type) {
    case IRS_JURISDICTIONS_FETCHED:
      if (action.IRSJurisdictionsById) {
        return SeamlessImmutable({
          ...state,
          IRSJurisdictionsById: { ...state.IRSJurisdictionsById, ...action.IRSJurisdictionsById },
        });
      }
      return state;
    case ADD_IRS_JURISDICTION:
      if (action.obj as IRSJurisdiction) {
        return SeamlessImmutable({
          ...state,
          IRSJurisdictionsById: {
            ...state.IRSJurisdictionsById,
            [action.obj.id as string]: action.obj,
          },
        });
      }
      return state;
    case REMOVE_IRS_JURISDICTIONS:
      return SeamlessImmutable({
        ...state,
        IRSJurisdictionsById: action.IRSJurisdictionsById,
      });
    default:
      return state;
  }
}

// selectors

/** get IRSJurisdictions by id
 * @param {Partial<Store>} state - the redux store
 * @param {string} planId - the plan id
 * @param {string} parentId - the parent id
 * @returns {{ [key: string]: IRSJurisdiction }} IRSJurisdictions by id
 */
export function getIRSJurisdictionsById(
  state: Partial<Store>,
  planId: string | null = null,
  parentId: string | null = null
): { [key: string]: IRSJurisdiction } {
  return keyBy(getIRSJurisdictionsArray(state, planId, parentId), 'id');
}

/** get one IRSJurisdiction using its id
 * @param {Partial<Store>} state - the redux store
 * @param {string} id - the IRSJurisdiction id
 * @returns {IRSJurisdiction|null} a IRSJurisdiction object or null
 */
export function getIRSJurisdictionById(
  state: Partial<Store>,
  id: string
): NamibiaIRSJurisdiction | null {
  return get((state as any)[reducerName].IRSJurisdictionsById, id) || null;
}

/** get an array of IRSJurisdiction objects
 * @param {Partial<Store>} state - the redux store
 * @param {string} planId - the plan id
 * @param {string} parentId - the parent id
 * @returns {IRSJurisdiction[]} an array of IRSJurisdiction objects
 */
export function getIRSJurisdictionsArray(
  state: Partial<Store>,
  planId: string | null = null,
  parentId: string | null = null
): IRSJurisdiction[] {
  let result = values((state as any)[reducerName].IRSJurisdictionsById);
  if (planId) {
    result = result.filter((e: IRSJurisdiction) => e.plan_id === planId);
  }
  if (parentId) {
    return result.filter((e: IRSJurisdiction) => e.jurisdiction_parent_id === parentId);
  }
  return result;
}
