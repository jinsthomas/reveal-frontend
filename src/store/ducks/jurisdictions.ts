import { FlexObject } from '@onaio/drill-down-table/dist/types/helpers/utils';
import { get, keyBy, keys, pickBy, values } from 'lodash';
import { AnyAction, Store } from 'redux';
import SeamlessImmutable from 'seamless-immutable';
import { GeoJSON, Geometry } from '../../helpers/utils';
import store from '../../store';

export const reducerName = 'jurisdictions';

/** interface for jurisdiction GeoJSON */
export interface JurisdictionGeoJSON extends GeoJSON {
  geometry: Geometry;
  properties: {
    jurisdiction_name: string;
    jurisdiction_parent_id: string;
  };
}

// todo - distingquish non-geo Jurisdiction, inheritted by new JurisdictionGeo type
/** interface to describe Jurisdiction */
export interface Jurisdiction {
  geographic_level?: number;
  geojson?: JurisdictionGeoJSON;
  jurisdiction_id: string;
  name?: string | null;
  parent_id?: string | null;
}

export interface AllJurisdictionIds {
  [key: string]: {
    id: string;
    isLoaded: boolean;
  };
}

// actions
export const FETCH_JURISDICTION = 'reveal/reducer/jurisdiction/FETCH_JURISDICTION';
export const FETCH_ALL_JURISDICTION_IDS = 'reveal/reducer/jurisdiction/FETCH_ALL_JURISDICTION_IDS';

/** fetch jurisdiction action */
interface FetchJurisdictionAction extends AnyAction {
  allJurisdictionIds?: AllJurisdictionIds;
  jurisdictionsById: { [key: string]: Jurisdiction };
  type: typeof FETCH_JURISDICTION;
}

interface FetchAllJurisdictionIdsAction extends AnyAction {
  allJurisdictionIds: AllJurisdictionIds;
  type: typeof FETCH_ALL_JURISDICTION_IDS;
}

/** jurisdiction action types */
export type JurisdictionActionTypes =
  | FetchJurisdictionAction
  | FetchAllJurisdictionIdsAction
  | AnyAction;

/** interface to describe jurisdiction state */
interface JurisdictionState {
  allJurisdictionIds: AllJurisdictionIds;
  jurisdictionsById: { [key: string]: Jurisdiction };
}

/** immutable Jurisdiction state */
export type ImmutableJurisdictionState = JurisdictionState &
  SeamlessImmutable.ImmutableObject<JurisdictionState>;

/** initial state */
const initialState: ImmutableJurisdictionState = SeamlessImmutable({
  allJurisdictionIds: {} as AllJurisdictionIds,
  jurisdictionsById: {},
});

// reducer
/** jurisdiction reducer function */
export default function reducer(
  state = initialState,
  action: JurisdictionActionTypes
): ImmutableJurisdictionState {
  switch (action.type) {
    case FETCH_JURISDICTION:
      if (action.jurisdictionsById) {
        return SeamlessImmutable({
          ...state,
          allJurisdictionIds: action.allJurisdictionIds,
          jurisdictionsById: {
            ...state.jurisdictionsById,
            ...action.jurisdictionsById,
          },
        });
      }
      return state;
    case FETCH_ALL_JURISDICTION_IDS:
      return SeamlessImmutable({
        ...state,
        allJurisdictionIds: action.allJurisdictionIds,
      });
    default:
      return state;
  }
}

// action creators
/** fetch Jurisdiction creator
 * @param {Jurisdiction[]} jurisdictionList - array of jurisdiction objects
 * @returns {FetchJurisdictionAction} FetchJurisdictionAction
 */
export const fetchJurisdictions = (jurisdictionList: Jurisdiction[] = []) => {
  return {
    allJurisdictionIds: keyBy(
      jurisdictionList.map((j: Jurisdiction) => ({
        id: j.jurisdiction_id,
        isLoaded: typeof j.geojson !== 'undefined',
      })),
      j => j.id
    ),
    jurisdictionsById: keyBy(
      jurisdictionList.map((item: Jurisdiction) => {
        const previousItem = getJurisdictionById(store.getState(), item.jurisdiction_id);
        /** ensure geojson is parsed */
        if (typeof item.geojson === 'string') {
          item.geojson = JSON.parse(item.geojson);
        }
        /** ensure geometry is parsed */
        if (item.geojson && typeof item.geojson.geometry === 'string') {
          item.geojson.geometry = JSON.parse(item.geojson.geometry);
        }
        if (previousItem) {
          return {
            ...previousItem,
            ...item,
          };
        }
        return item;
      }),
      jurisdiction => jurisdiction.jurisdiction_id
    ),
    type: FETCH_JURISDICTION,
  };
};

/** fetch all Jurisdiction Ids creator */
export const fetchAllJurisdictionIds = (jurisdictionIds: string[]) => ({
  allJurisdictionIds: keyBy(
    jurisdictionIds.map((id: string) => ({
      id,
      isLoaded: !!(
        getJurisdictionById(store.getState(), id) &&
        (getJurisdictionById(store.getState(), id) as Jurisdiction).geojson
      ),
    })),
    j => j.id
  ),
  type: FETCH_ALL_JURISDICTION_IDS,
});

// selectors
/** get jurisdictions by id
 * @param {Partial<Store>} state - the redux store
 */
export function getJurisdictionsById(state: Partial<Store>): { [key: string]: Jurisdiction } {
  return (state as any)[reducerName].jurisdictionsById;
}

/** get one jurisdiction using its id
 * @param {Partial<Store>} state - the redux store
 * @param {string} id - the jurisdiction id
 * @returns {Jurisdiction|null} a jurisdiction or null
 */
export function getJurisdictionById(state: Partial<Store>, id: string): Jurisdiction | null {
  return get((state as any)[reducerName].jurisdictionsById, id) || null;
}

/** get an array of jurisdiction objects
 * @param {Partial<Store>} state - the redux store
 * @returns {Jurisdiction[]} an array of jurisdictions
 */
export function getJurisdictionsArray(state: Partial<Store>): Jurisdiction[] {
  return values((state as any)[reducerName].jurisdictionsById);
}

/** get an array of jurisdiction ids
 * @param {Partial<Store>} state - the redux store
 * @returns {string[]} an array of jurisdictions ids
 */
export function getJurisdictionsIdArray(state: Partial<Store>): string[] {
  return keys((state as any)[reducerName].jurisdictionsById);
}

export function getAllJurisdictionsIds(state: Partial<Store>): string[] {
  return (state as any)[reducerName].allJurisdictionIds;
}

/** get an array of all jurisdiction ids
 * @param {Partial<Store>} state - the redux store
 * @param {boolean|undefined} isLoaded optional - filter by isLoaded
 * @returns an array of all jurisdictions available from the server
 */
export function getAllJurisdictionsIdArray(
  state: Partial<Store>,
  isLoaded?: boolean | undefined
): string[] {
  if (typeof isLoaded === 'undefined') {
    return keys((state as any)[reducerName].allJurisdictionIds);
  }
  return keys(
    pickBy((state as any)[reducerName].allJurisdictionIds, (j: FlexObject) =>
      isLoaded ? j.isLoaded : !j.isLoaded
    )
  );
}
