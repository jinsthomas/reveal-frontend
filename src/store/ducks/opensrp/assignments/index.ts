/** Assignments redux module */
import { get } from 'lodash';
import { Store } from 'redux';
import { AnyAction } from 'redux';
import SeamlessImmutable from 'seamless-immutable';

/** The reducer name */
export const reducerName = 'assignments';

/** interface for a Assignment object */
export interface Assignment {
  jurisdiction: string;
  organization: string;
  plan: string;
  fromDate: string;
  toDate: string;
}

// action interfaces

/** action type for action that adds Assignments to store */
export const ASSIGNMENTS_FETCHED = 'src/store/ducks/assignments/reducer/ASSIGNMENTS_FETCHED';
/** action type for REMOVE_TEAMS action */
export const REMOVE_ASSIGNMENTS = 'src/store/ducks/assignments/reducer/REMOVE_ASSIGNMENTS';

/** interface for Assignments fetched action */
interface FetchAssignmentsAction extends AnyAction {
  assignmentsByPlanId: { [key: string]: Assignment[] };
  type: typeof ASSIGNMENTS_FETCHED;
}

/** interface for action that removes assignments from store */
interface RemoveAssignmentsAction extends AnyAction {
  assignmentsByPlanId: { [key: string]: Assignment[] };
  type: typeof REMOVE_ASSIGNMENTS;
}

/** single type for all action types */
type AssignmentActionTypes = FetchAssignmentsAction | RemoveAssignmentsAction | AnyAction;

/** interface for Assignments state in store */
interface AssignmentsStoreState {
  assignmentsByPlanId: { [key: string]: Assignment[] };
}

// immutable assignments state in dux
export type ImmutableAssignmentsStoreState = AssignmentsStoreState &
  SeamlessImmutable.ImmutableObject<AssignmentsStoreState>;

/** initial state for Assignments records in store */
const initialOrgsStoreState: ImmutableAssignmentsStoreState = SeamlessImmutable({
  assignmentsByPlanId: {},
});

/** the Assignment reducer function */
export default function reducer(state = initialOrgsStoreState, action: AssignmentActionTypes) {
  switch (action.type) {
    case ASSIGNMENTS_FETCHED:
      return SeamlessImmutable({
        ...state,
        assignmentsByPlanId: {
          ...state.assignmentsByPlanId,
          ...action.assignmentsByPlanId,
        },
      });
    case REMOVE_ASSIGNMENTS:
      return SeamlessImmutable({
        ...state,
        assignmentsByPlanId: action.assignmentsByPlanId,
      });
    default:
      return state;
  }
}

/** action to remove assignments form store */
export const removeAssignmentsAction: RemoveAssignmentsAction = {
  assignmentsByPlanId: {},
  type: REMOVE_ASSIGNMENTS,
};

// action creators

/** creates action to add fetched assignments to store
 * @param {Assignment []} assignmentsList - array of assignments to be added to store
 *
 * @returns {FetchAssignmentsAction} - action with assignments payload that is added to store
 */
export const FetchAssignments = (assignmentsList: Assignment[]): FetchAssignmentsAction => {
  const assignmentsByPlanId: { [key: string]: Assignment[] } = {};

  for (const assignment of assignmentsList) {
    if (!assignmentsByPlanId[assignment.plan]) {
      assignmentsByPlanId[assignment.plan] = [];
    }
    assignmentsByPlanId[assignment.plan].push(assignment);
  }

  return {
    assignmentsByPlanId,
    type: ASSIGNMENTS_FETCHED,
  };
};

// selectors

/** get assignments as an object where their ids are the keys and the objects
 * the values
 * @param {Partial<Store>} state - Portion of the store
 *
 * @return {[key: string]: Assignment}
 */
export function getAssignmentsByPlanId(state: Partial<Store>): { [key: string]: Assignment[] } {
  return (state as any)[reducerName].assignmentsByPlanId;
}

/** Get all assignments as an array
 * @param {Partial<Store>} state - Part of the redux store
 * @param {string} planId - The id of the plan of which to return associated Assignments
 *
 * @return {Assignment []} - all assignments by planId in store as an array
 */
export function getAssignmentsArrayByPlanId(state: Partial<Store>, planId: string): Assignment[] {
  return get(getAssignmentsByPlanId(state), planId);
}
