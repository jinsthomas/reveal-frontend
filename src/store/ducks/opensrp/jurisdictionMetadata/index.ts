/** interface for JurisdictionMetadataImport object */
export interface JurisdictionMetadataImport {
  file: File;
}
