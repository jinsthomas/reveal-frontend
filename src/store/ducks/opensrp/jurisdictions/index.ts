import { Dictionary } from '@onaio/utils/dist/types/types';
import {
  fetchActionCreatorFactory,
  reducerFactory,
  removeActionCreatorFactory,
} from '@opensrp/reducer-factory';
import { Feature, FeatureCollection, Geometry } from '@turf/turf';
import { get, values } from 'lodash';
import { Store } from 'redux';
import { createSelector } from 'reselect';

/** The shape of a jurisdiction received from the OpenSRP API */
export interface Jurisdiction {
  id: string;
  geometry?: Geometry;
  properties: {
    code?: string;
    geographicLevel: number;
    name: string;
    parentId?: string;
    status: string;
    version: string | number;
  };
  serverVersion: number;
  type: Readonly<'Feature'>;
}

/** the reducer name */
export const reducerName = 'Jurisdiction';

/** Jurisdiction Reducer */
const reducer = reducerFactory<Jurisdiction>(reducerName);
export default reducer;

// actions
/** actionCreator returns action to to add Item records to store */
export const fetchJurisdictions = fetchActionCreatorFactory<Jurisdiction>(reducerName, 'id');
export const removeJurisdictions = removeActionCreatorFactory(reducerName);

// selectors
/** prop filters to customize selector queries */
export interface Filters {
  jurisdictionId?: string /** jurisdiction id */;
  jurisdictionIdsArray?: string[] /** array of jurisdiction ids */;
  parentId?: string /** parent id */;
}

/** retrieve the jurisdictionId value
 * @param state - the store
 * @param props -  the filterProps
 */
export const getJurisdictionId = (_: Partial<Store>, props: Filters) => props.jurisdictionId;

/** retrieve the parentId value
 * @param state - the store
 * @param props -  the filterProps
 */
export const getParentId = (_: Partial<Store>, props: Filters) => props.parentId;

/** retrieve the jurisdictionIdsArray value
 * @param state - the store
 * @param props -  the filterProps
 */
export const getJurisdictionIdsArray = (_: Partial<Store>, props: Filters) =>
  props.jurisdictionIdsArray;

/** gets all jurisdictions keyed by id
 * @param state - the store
 * @param _ -  the filterProps
 */
export const getJurisdictionsById = (state: Partial<Store>, _: Filters): Dictionary<Jurisdiction> =>
  (state as any)[reducerName].objectsById;

/** retrieve the Jurisdiction using an id
 * @param state - the store
 * @param props -  the filterProps
 */
export const getJurisdictionById = () =>
  createSelector(
    [getJurisdictionsById, getJurisdictionId],
    (jurisdictionsById, jurisdictionId): Jurisdiction | null => {
      if (!jurisdictionId) {
        return null;
      }
      return get(jurisdictionsById, jurisdictionId, null);
    }
  );

/**
 * Get jurisdiction ids
 * @param state - the store
 * @param props -  the filterProps
 */
export const getJurisdictionIds = () =>
  createSelector([getJurisdictionsById], (jurisdictionsById): string[] => {
    return Object.keys(jurisdictionsById);
  });

/**
 * retrieve jurisdictions as an array
 * @param state - the store
 * @param props -  the filterProps
 */
export const getJurisdictionsArray = () =>
  createSelector(
    [getJurisdictionsById, getParentId, getJurisdictionIdsArray],
    (jurisdictionsById, parentId, jurisdictionIdsArray): Jurisdiction[] => {
      const result = values(jurisdictionsById);
      if (parentId) {
        return result.filter(item => item.properties.parentId === parentId);
      }
      if (jurisdictionIdsArray) {
        return result.filter(item => jurisdictionIdsArray.includes(item.id));
      }
      return result;
    }
  );

/**
 * retrieve jurisdictions as a geojson feature collection
 * @param state - the store
 * @param props -  the filterProps
 */
export const getJurisdictionsFC = () =>
  createSelector(
    [getJurisdictionsArray()],
    (jurisdictionsArray): FeatureCollection => {
      return {
        features: jurisdictionsArray.filter(item => 'geometry' in item) as Feature[],
        type: 'FeatureCollection',
      };
    }
  );
