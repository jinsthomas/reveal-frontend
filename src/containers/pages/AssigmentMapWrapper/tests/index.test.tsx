import React from 'react';
import { Provider } from 'react-redux';
import { Router } from 'react-router';
import { mount, shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import flushPromises from 'flush-promises';
import { createBrowserHistory } from 'history';
import { AssignmentMapWrapper } from '..';
import * as jurisdictionDucks from '../../../../store/ducks/opensrp/jurisdictions';
import * as fixtures from './fixtures';
import { FeatureCollection } from '@turf/turf';

jest.mock('../../../../components/GisidaLite', () => {
  const GisidaLiteMock = () => <div>I love oov</div>;
  return {
    GisidaLite: GisidaLiteMock,
  };
});
jest.mock('../../../../configs/env');
const history = createBrowserHistory();

describe('containers/pages/AssigmentMapWrapper', () => {
  it('renders successfully', () => {
    shallow(
      <Router history={history}>
        <AssignmentMapWrapper />
      </Router>
    );
  });
  it('renders jurisdictions assignment map', () => {
    const props = {
      currentParentId: '07b09ec1-0589-4a98-9480-4c403ac24d59',
      rootJurisdictionId: '872cc59e-0bce-427a-bd1f-6ef674dba8e2',
      getJurisdictionsFeatures: fixtures.geoCollection as FeatureCollection,
    };
    const wrapper = mount(
      <Router history={history}>
        <AssignmentMapWrapper {...props} />
      </Router>
    );
    expect(toJson(wrapper.find('.map'))).toMatchSnapshot();
    wrapper.unmount();
  });
});
