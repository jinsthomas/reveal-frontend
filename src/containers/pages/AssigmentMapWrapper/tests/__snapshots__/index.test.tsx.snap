// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`containers/pages/AssigmentMapWrapper renders jurisdictions assignment map 1`] = `null`;
